### DISCUSSION TITLE
Does anybody know how to convert a 1D array to a std::vector in the most optimal way?

### DISCUSSION THREAD

**Discussion by cemlyn007** [ml-explore/mlx#773] [Q&A] [answered]
I know we can do this?
```
std::vector<int> snake_vector = {};
for (auto value: snake) {
    snake_vector.push_back(i);
}
```

But I am wondering if anybody has had any success doing something faster? Maybe with the array data method?

I want to essentially want to extract all non-zero elements from the array and this is an intermediate step to do it

**Comment by angeloskath** [comment, marked as answer]
This is all about memory management. Since the vector manages its own memory then we can't just create a normal vector from an MLX array. If you just want to access the underlying elements though then it is as simple as 

```cpp
auto data = snake.data<int>();
for (int el = 0; el < snake.size(); el++) {
  // you can use elem_to_loc here or just el depending
  // on whether the data is contiguous or not
  //
  // do something with data[el]
}
```

**Reply by cemlyn007** [reply]
Dayum I am a noob, I was writing loads of mlx::core::takes earlier in a for-loop to do the access which was so slow! Thank you so much!
